packages/contracts-bedrock/src/L1/OPContractsManager.sol

```diff
@@ -668,263 +668,77 @@ contract OPContractsManagerUpgrader is OPContractsManagerBase {
-            // Separate context to avoid stack too deep.
-            {
-                // Now retrieve the permissionless game.
-                IFaultDisputeGame permissionlessDisputeGame =
-                    IFaultDisputeGame(address(getGameImplementation(dgf, GameTypes.CANNON)));
-
-                // If it exists, replace its implementation.
-                if (address(permissionlessDisputeGame) != address(0)) {
-                    // Create a new DelayedWETH for the permissionless game.
-                    IDelayedWETH permissionlessDelayedWeth = IDelayedWETH(
-                        payable(
-                            deployProxy({
-                                _l2ChainId: l2ChainId,
-                                _proxyAdmin: _opChainConfigs[i].proxyAdmin,
-                                _saltMixer: reusableSaltMixer(_opChainConfigs[i]),
-                                _contractName: "PermissionlessDelayedWETH-U16a"
-                            })
-                        )
-                    );
-
-                    // Initialize the DelayedWETH.
-                    upgradeToAndCall(
-                        _opChainConfigs[i].proxyAdmin,
-                        address(permissionlessDelayedWeth),
-                        impls.delayedWETHImpl,
-                        abi.encodeCall(IDelayedWETH.initialize, (_opChainConfigs[i].systemConfigProxy))
-                    );
-
-                    // Fix for stack too deep
-                    OPContractsManager.OpChainConfig memory opChainConfig = _opChainConfigs[i];
-                    // Deploy and set a new permissionless game to update its prestate
-                    deployAndSetNewGameImpl({
-                        _l2ChainId: l2ChainId,
-                        _disputeGame: IDisputeGame(address(permissionlessDisputeGame)),
-                        _newDelayedWeth: permissionlessDelayedWeth,
-                        _newAnchorStateRegistryProxy: anchorStateRegistry,
-                        _gameType: GameTypes.CANNON,
-                        _opChainConfig: opChainConfig
-                    });
-                }
-            }
+            // Do the chain upgrade.
+            // All of your updates should be done in this internal function unless you're making a
+            // change to how upgrades work in general.
+            _doChainUpgrade(impls, _opChainConfigs[i], l2ChainId);
 
             // Emit the upgraded event with the address of the caller. Since this will be a delegatecall,
             // the caller will be the value of the ADDRESS opcode.
             emit Upgraded(l2ChainId, _opChainConfigs[i].systemConfigProxy, address(this));
         }
     }
 
+    /// @notice Performs an upgrade for a specific chain.
+    /// @param _impls The implementations of the contracts.
+    /// @param _opChainConfig The configuration of the chain to upgrade.
+    /// @param _l2ChainId The L2 chain ID of the chain to upgrade.
+    function _doChainUpgrade(
```

Comment:
I've created a new function here so that it's simpler for 3rd party users of the OPCM to know what to touch and what not to touch.